File: src/com/owncloud/android/ui/activity/ManageSpaceActivity.java
package com.owncloud.android.ui.activity;

import android.app.Activity;
import android.app.ActivityManager;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.owncloud.android.R;
import com.owncloud.android.lib.common.utils.Log_OC;

import java.io.File;

public class ManageSpaceActivity extends AppCompatActivity {

    private static final String TAG = ManageSpaceActivity.class.getSimpleName();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_manage_space);

        ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setTitle(R.string.manage_space_title);

        TextView descriptionTextView = (TextView) findViewById(R.id.general_description);
        descriptionTextView.setText(getString(R.string.manage_space_description, getString(R.string.app_name)));

        Button clearDataButton = (Button) findViewById(R.id.clearDataButton);
        clearDataButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                clearData();
            }
        });
    }

    /**
     * Save passcode from Share preferences
     * Clear the rest of data
     */
    private void clearData() {
        // Save passcode from Share preferences
        SharedPreferences appPrefs = PreferenceManager
                .getDefaultSharedPreferences(getApplicationContext());

        boolean passCodeEnable = appPrefs.getBoolean("set_pincode", false);

        String passCodeDigits[] = new String[4];
        if (passCodeEnable) {
            passCodeDigits[0] = appPrefs.getString("PrefPinCode1", null);
            passCodeDigits[1] = appPrefs.getString("PrefPinCode2", null);
            passCodeDigits[2] = appPrefs.getString("PrefPinCode3", null);
            passCodeDigits[3] = appPrefs.getString("PrefPinCode4", null);
        }

        // Clear data
        clearApplicationData();

        // Recover passcode
        SharedPreferences.Editor appPrefsEditor = PreferenceManager
                .getDefaultSharedPreferences(getApplicationContext()).edit();
        if (passCodeEnable) {
            appPrefsEditor.putString("PrefPinCode1", passCodeDigits[0]);
            appPrefsEditor.putString("PrefPinCode2", passCodeDigits[1]);
            appPrefsEditor.putString("PrefPinCode3", passCodeDigits[2]);
            appPrefsEditor.putString("PrefPinCode4", passCodeDigits[3]);
        }

        appPrefsEditor.putBoolean("set_pincode", passCodeEnable);
        appPrefsEditor.commit();


        finish();
        System.exit(0);

    }

    public void clearApplicationData() {
        File cache = getCacheDir();
        File appDir = new File(cache.getParent());
        if (appDir.exists()) {
            String[] children = appDir.list();
            for (String s : children) {
                if (!s.equals("lib")) {
                    File fileToDelete = new File(appDir, s);
                    deleteDir(fileToDelete);
                    Log_OC.d(TAG, "Clear Application Data, File: " + fileToDelete.getName()+ " DELETED *******");
                }
            }
        }
    }

    public static boolean deleteDir(File dir) {
        if (dir != null && dir.isDirectory()) {
            String[] children = dir.list();
            for (int i = 0; i < children.length; i++) {
                boolean success = deleteDir(new File(dir, children[i]));
                if (!success) {
                    return false;
                }
            }
        }

        return dir.delete();
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        super.onOptionsItemSelected(item);
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
            default:
                Log_OC.w(TAG, "Unknown menu item triggered");
                return false;
        }
        return true;
    }
}

File: app/src/main/java/com/owncloud/android/ui/activity/ManageSpaceActivity.kt
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2025 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-FileCopyrightText: 2022 Álvaro Brey <alvaro@alvarobrey.com>
 * SPDX-License-Identifier: AGPL-3.0-or-later OR GPL-2.0-only
 */
package com.owncloud.android.ui.activity

import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.google.android.material.snackbar.Snackbar
import com.nextcloud.android.common.ui.util.extensions.applyEdgeToEdgeWithSystemBarPadding
import com.nextcloud.client.account.UserAccountManager
import com.nextcloud.client.di.Injectable
import com.nextcloud.client.preferences.AppPreferences
import com.owncloud.android.R
import com.owncloud.android.databinding.ActivityManageSpaceBinding
import com.owncloud.android.lib.common.utils.Log_OC
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import javax.inject.Inject
import kotlin.system.exitProcess

class ManageSpaceActivity :
    AppCompatActivity(),
    Injectable {

    @Inject
    lateinit var preferences: AppPreferences

    @Inject
    lateinit var userAccountManager: UserAccountManager

    private lateinit var binding: ActivityManageSpaceBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        applyEdgeToEdgeWithSystemBarPadding()
        super.onCreate(savedInstanceState)

        binding = ActivityManageSpaceBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.run {
            manageActivityToolbar.setNavigationOnClickListener {
                onBackPressedDispatcher.onBackPressed()
            }
            generalDescription.text = getString(R.string.manage_space_description, getString(R.string.app_name))
            clearDataButton.setOnClickListener {
                lifecycleScope.launch {
                    clearData()
                }
            }
        }
    }

    @Suppress("MagicNumber")
    private suspend fun clearData() {
        withContext(Dispatchers.IO) {
            val lockPref = preferences.lockPreference
            val passCodeEnable = SettingsActivity.LOCK_PASSCODE == lockPref
            var passCodeDigits = arrayOfNulls<String>(4)
            if (passCodeEnable) {
                passCodeDigits = preferences.passCode
            }

            // Clear preferences data
            preferences.clear()

            // Recover passcode
            if (passCodeEnable) {
                preferences.setPassCode(
                    passCodeDigits[0],
                    passCodeDigits[1],
                    passCodeDigits[2],
                    passCodeDigits[3]
                )
            }
            preferences.lockPreference = lockPref
            userAccountManager.removeAllAccounts()

            // Clear app data
            val result = clearApplicationData()
            withContext(Dispatchers.Main) {
                if (result) {
                    finishAndRemoveTask()
                    exitProcess(0)
                } else {
                    Snackbar.make(
                        findViewById(android.R.id.content),
                        R.string.manage_space_clear_data,
                        Snackbar.LENGTH_LONG
                    ).show()
                }
            }
        }
    }

    @Suppress("NestedBlockDepth")
    private fun clearApplicationData(): Boolean {
        var clearResult = true

        cacheDir.parent?.let { parentCacheDirPath ->
            val appDir = File(parentCacheDirPath)
            if (appDir.exists()) {
                val children = appDir.list()
                if (children != null) {
                    children.filter { it != LIB_FOLDER }.forEach { s ->
                        val fileToDelete = File(appDir, s)
                        clearResult = clearResult && deleteDir(fileToDelete)
                        Log_OC.d(TAG, "Clear Application Data, File: " + fileToDelete.name + " DELETED *****")
                    }
                } else {
                    clearResult = false
                }
            }
        }

        return clearResult
    }

    @Suppress("ReturnCount")
    private fun deleteDir(dir: File?): Boolean {
        if (dir != null && dir.isDirectory) {
            dir.list()?.forEach { child ->
                val success = deleteDir(File(dir, child))
                if (!success) {
                    Log_OC.w(TAG, "File NOT deleted $child")
                    return false
                } else {
                    Log_OC.d(TAG, "File deleted $child")
                }
            } ?: return false
        }
        return dir?.delete() ?: false
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        android.R.id.home -> {
            finish()
            true
        }
        else -> {
            Log_OC.w(TAG, "Unknown menu item triggered")
            super.onOptionsItemSelected(item)
        }
    }

    companion object {
        private val TAG = ManageSpaceActivity::class.java.simpleName
        private const val LIB_FOLDER = "lib"
    }
}
